bcm53xx: add testing kernel 6.12

stklcode opened the pull request:
Straightforward kernel bump.

Drop upstreamed patches:
* 030-v6.7-0001-ARM-dts-BCM5301X-Set-MACs-for-D-Link-DIR-885L.patch
* 030-v6.7-0002-ARM-dts-BCM5301X-Set-MAC-address-for-Asus-RT-AC87U.patch
* 030-v6.7-0003-ARM-dts-BCM5301X-Relicense-Felix-s-code-to-the-GPL-2.patch
* 030-v6.7-0004-ARM-dts-BCM5301X-Relicense-Vivek-s-code-to-the-GPL-2.patch
* 030-v6.7-0005-ARM-dts-BCM5301X-Explicitly-disable-unused-switch-CP.patch
* 030-v6.7-0006-ARM-dts-BCM5301X-Set-fixed-link-for-extra-Netgear-R8.patch
* 030-v6.7-0007-ARM-dts-BCM5301X-Set-switch-ports-for-Linksys-EA9200.patch
* 040-v6.11-ARM-dts-broadcom-convert-NVMEM-content-to-layout-syntax.patch

All other patches and configs automatically refreshed.

Compile tested only. I do not have any compatible device for run-testing.

- - - - 

Run tests:

* Linksys EA9200 ❌ (@raenye) - bootloop, partitioning failure
* Linksys EA9500 ✅  (@Stephen2301)
* Asus RT-N18U ✅ (@strtrt)
* Asus RT-AC68U ✅ (cookiemonster)
* Meraki MX65 ✅ (@DimmaDont)

stklcode committed:
bcm53xx: 6.12: refresh patches

Drop upstreamed patches:
030-v6.7-0001-ARM-dts-BCM5301X-Set-MACs-for-D-Link-DIR-885L.patch
030-v6.7-0002-ARM-dts-BCM5301X-Set-MAC-address-for-Asus-RT-AC87U.patch
030-v6.7-0003-ARM-dts-BCM5301X-Relicense-Felix-s-code-to-the-GPL-2.patch
030-v6.7-0004-ARM-dts-BCM5301X-Relicense-Vivek-s-code-to-the-GPL-2.patch
030-v6.7-0005-ARM-dts-BCM5301X-Explicitly-disable-unused-switch-CP.patch
030-v6.7-0006-ARM-dts-BCM5301X-Set-fixed-link-for-extra-Netgear-R8.patch
030-v6.7-0007-ARM-dts-BCM5301X-Set-switch-ports-for-Linksys-EA9200.patch
040-v6.11-ARM-dts-broadcom-convert-NVMEM-content-to-layout-syntax.patch

All other patches automatically refreshed.

Signed-off-by: Stefan Kalscheuer <stefan@stklcode.de>
Link: https://github.com/openwrt/openwrt/pull/20167
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

stklcode committed:
bcm53xx: 6.12: refresh kernel config

Add new and remove obsolete symbols for Kernel 6.12 by running
make kernel_oldconfig.

Signed-off-by: Stefan Kalscheuer <stefan@stklcode.de>
Link: https://github.com/openwrt/openwrt/pull/20167
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

stklcode committed:
bcm53xx: add 6.12 testing kernel

Add 6.12 testing kernel for bcm53xx target.

Signed-off-by: Stefan Kalscheuer <stefan@stklcode.de>
Link: https://github.com/openwrt/openwrt/pull/20167
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

linusw reviewed (approved):
Looks good to me.

stklcode commented:
Hm, sounds like it resolves the issue. I’d be happy to drop it for good, in a separate change/commit before or after the kernel bump.

Unfortunately I neither own a device that required the patch nor one that has the linked issue, so I have to rely on user feedback here.

raenye commented:
efd4445 (= PR 20362) has just been merged, maybe have it for 6.12 too?

hauke commented:
Was this run tested by anyone? If you have such a device please test this PR and report back.

Stephen2301 commented:
Both usb2 and usb3 are working

GoetzGoerisch commented:
build-tested on top of 6.12.54

Ansuel commented:
@stklcode I'm a bit confused what is the problem with this? USB still broken? Some device bootloops?

raenye commented:
@Ansuel My Linksys EA9200 bootloops on 6.12, please see [above](https://github.com/openwrt/openwrt/pull/20167#issuecomment-3421430702).
USB3 (xhci) has been broken since Kernel 6.1, while USB2 (ohci) works fine.

Ansuel commented:
I think there is a problem in parsing the TRX partition... Maybe there is an extra check there that is causing -22 ? We should bisect TRX parser.

stklcode commented:
> Is TRX parser a part of the kernel? I'm confused.

Yes. TRX is the partition format for various Broadcom boards and part of the kernel's mtd driver:
* https://www.kernel.org/doc/Documentation/devicetree/bindings/mtd/partitions/brcm%2Ctrx.txt
* https://github.com/gregkh/linux/blob/v6.12.55/drivers/mtd/parsers/parser_trx.c

The EA9200 has `compatible = "brcm,trx"` partitions that apparently fail to initialize.
Can't be entirely broken though, as the EA9500, which was successfully tested, also has TRX partitions.

raenye commented:
> @raenye can you try [0001-Revert-mtd-mtdpart-check-for-subpartitions-parsing-r.patch](https://github.com/user-attachments/files/23360540/0001-Revert-mtd-mtdpart-check-for-subpartitions-parsing-r.patch)

Works now! thanks.

```
...
[    1.916700] 6 fixed-partitions partitions found on MTD device brcmnand.0
[    1.923847] Creating 6 MTD partitions on "brcmnand.0":
[    1.929047] 0x000000000000-0x000000080000 : "boot"
[    1.934517] 0x000000080000-0x000000180000 : "nvram"
[    1.940127] 0x000000180000-0x000000200000 : "devinfo"
[    1.945858] 0x000000200000-0x000002a00000 : "backup"
[    1.951680] 0x000002a00000-0x000005200000 : "firmware"
[    1.958256] 2 trx partitions found on MTD device firmware
[    1.963686] Creating 2 MTD partitions on "firmware":
[    1.968707] 0x00000000001c-0x000000400000 : "linux"
[    1.973605] mtd: partition "linux" doesn't start on an erase/write block boundary -- force read-only
[    1.983178] 0x000000400000-0x000002800000 : "ubi"
...
```

Ansuel commented:
@Stephen2301 can I have your boot log please?

Ansuel commented:
@raenye news?

Ansuel commented:
@raenye ok i feel stupid now... i was always checking the upstream trx and was always confused since the driver doesn't have any -EINVAL error (-22)... I just now checked if we make downstream change to it and...... patch 140..... it was in plain sight........

Ansuel commented:
Ok took me a solid 10 minutes to understand why 140 is needed but now it's clear.

the scenario is the following.
2 trx partition that both expose kernel and ubi partition.

In such case we end up with 2 kernel and 2 ubi partition and this can be problematic.

The solution is to skip parsing them so we have a single ubi and kernel partition.

raenye commented:
If 140 is the only reason for EINVAL, then the check for negative return value could just do something like

if return value is EINVAL, print "TRX parsing of failsafe partition skipped", otherwise goto delete_partitions_stuff.

Ansuel commented:
guys chill, I got all the information and I know what solution to apply. no need to give suggestion for solution.

I will push upstream a change to skip classifying -ENODEV as an error in mtdpart.c (because it's a clear regression to take all error from a parser as "errors") and change the parser to return -ENODEV if something is wrong.

Even in the TRX parser there is a problem. In the context of the partition wiped, the trx_magic will be zero and an error will be printed. But a wiped TRX partition is totally ok and should be skipped (aka ignored as the old implementation where is a parser fails to parse the partition is just ignored instead of being an hard condition).

stklcode committed:
kernel/bcm53xx: Restore kernel files for v6.6

This is an automatically generated commit which aids following Kernel patch
history, as git will see the move and copy as a rename thus defeating the
purpose.

For the original discussion see:
https://lists.openwrt.org/pipermail/openwrt-devel/2023-October/041673.html

Signed-off-by: Stefan Kalscheuer <stefan@stklcode.de>
Link: https://github.com/openwrt/openwrt/pull/20167
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

stklcode committed:
kernel/bcm53xx: Create kernel files for v6.12 (from v6.6)

This is an automatically generated commit.

When doing `git bisect`, consider `git bisect --skip`.

Signed-off-by: Stefan Kalscheuer <stefan@stklcode.de>
Link: https://github.com/openwrt/openwrt/pull/20167
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
I will wait 2-3 more days for a feedback upstream... Then I will just backport the patch in pending and merge this.

DimmaDont commented:
Built and run tested on a Meraki MX65.
[mx65_minicom.log](https://github.com/user-attachments/files/23488091/mx65_minicom.log)

udanfacy22 commented:
Cookiemonster on the OpenWrt forums has run tested this on an RT-AC68U but has been unable to comment here due to getting incorrectly flagged as a bot. See their two posts here: https://forum.openwrt.org/t/kernel-6-12-debugging-and-testing-needed-for-at91-bcm47xx-bcm4908-bcm53xx-mpc85xx-qoriq-siflower-and-zynq/242857/14

stklcode commented:
@raenye Should be fine in the current state. Both fixes have been applied to _main_ before this PR was merged.

Fix for the TRX parser to return `~ENOENT` instead of `~EINVAL`: https://github.com/openwrt/openwrt/commit/231bbe528d3f6ae4a43117b38c41a2f62896ce4e
Backported upstream fix for the MTD parser to properly handle `~ENOENT`: https://github.com/openwrt/openwrt/commit/2434ad8b0815fea93aa05c1e444457cc340f9e8e

Self-Hosting-Group commented:
[See issue](https://redirect.github.com/openwrt/openwrt/issues/21187)
